File: app/Http/Controllers/PedidoController.php
<?php

namespace App\Http\Controllers;

use App\Models\Cliente;
use App\Models\DetallePedido;
use App\Models\Envio;
use App\Models\Pago;
use App\Models\Plato;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;

class PedidoController extends Controller
{
    public function confirmar(Request $request): JsonResponse
    {
        if (! session('CodigoUsuario')) {
            return response()->json(['error' => 'No autenticado', 'redirect' => '/login'], 401);
        }

        $items = $request->input('items', []);
        $metodoPago = $request->input('metodo_pago', 'Tarjeta');

        if (empty($items)) {
            return response()->json(['error' => 'El carrito está vacío.'], 422);
        }

        $cliente = Cliente::where('CodigoUsuario', session('CodigoUsuario'))->first();

        if (! $cliente) {
            return response()->json(['error' => 'No se encontró el perfil de cliente.'], 422);
        }

        $codigoRestaurante = $items[0]['restaurante_id'] ?? null;

        if (! $codigoRestaurante) {
            $plato = Plato::with('menu')->find($items[0]['id'] ?? null);
            $codigoRestaurante = $plato?->menu?->CodigoRestaurante;
        }

        if (! $codigoRestaurante) {
            return response()->json(['error' => 'No se pudo determinar el restaurante.'], 422);
        }

        $total = collect($items)->sum(fn ($item) => ($item['precio'] ?? 0) * ($item['cantidad'] ?? 1));

        try {
            $codigoPedido = null;

            DB::transaction(function () use ($cliente, $codigoRestaurante, $items, $total, &$codigoPedido) {
                $envio = Envio::create([
                    'CodigoCliente' => $cliente->CodigoCliente,
                    'CodigoRepartidor' => 1,
                    'CodigoRestaurante' => $codigoRestaurante,
                    'Descripcion' => 'Pedido web',
                    'FechaEnvio' => now()->toDateString(),
                    'HoraEntrega' => now()->addMinutes(35)->format('H:i:s'),
                ]);

                $pedido = $envio->pedido()->firstOrFail();
                $codigoPedido = $pedido->CodigoPedido;

                foreach ($items as $item) {
                    DetallePedido::create([
                        'CodigoPedido' => $pedido->CodigoPedido,
                        'CodigoPlato' => $item['id'],
                        'Cantidad' => (int) ($item['cantidad'] ?? 1),
                        'PrecioUnitario' => (float) ($item['precio'] ?? 0),
                    ]);
                }

                Pago::create([
                    'CodigoCliente' => $cliente->CodigoCliente,
                    'CodigoEnvio' => $envio->CodigoEnvio,
                    'Monto' => $total,
                    'FechaPago' => now()->toDateString(),
                    'HoraPago' => now()->format('H:i:s'),
                    'EstadoPago' => 'Aceptado',
                ]);

                session(['pedido_activo' => $codigoPedido]);
            });

            return response()->json(['redirect' => '/rastreo']);
        } catch (\Throwable $e) {
            return response()->json(['error' => 'Error al procesar el pedido. Inténtalo nuevamente.'], 500);
        }
    }
}

File: src/main/java/com/urbaneats/controller/PedidoController.java
package com.urbaneats.controller;

import com.urbaneats.dao.ClienteRepository;
import com.urbaneats.dao.PedidoRepository;
import com.urbaneats.entity.Cliente;
import com.urbaneats.entity.Pago;
import com.urbaneats.entity.Pedido;
import com.urbaneats.security.CustomUserDetails;
import com.urbaneats.service.PedidoDesdeCarritoService;
import jakarta.servlet.http.HttpSession;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.*;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;
import java.util.Map;

/**
 * Confirmacion de pedido sin pasarela de pago, consulta de estado y cancelacion.
 * Equivale a PedidoController de la rama php-Laravel (rutas bajo /pedido, no /pedidos).
 */
@Controller
public class PedidoController {

    private final ClienteRepository clienteRepository;
    private final PedidoRepository pedidoRepository;
    private final PedidoDesdeCarritoService pedidoDesdeCarritoService;

    public PedidoController(ClienteRepository clienteRepository,
                             PedidoRepository pedidoRepository,
                             PedidoDesdeCarritoService pedidoDesdeCarritoService) {
        this.clienteRepository = clienteRepository;
        this.pedidoRepository = pedidoRepository;
        this.pedidoDesdeCarritoService = pedidoDesdeCarritoService;
    }

    private CustomUserDetails getUsuarioActual() {
        Authentication auth = SecurityContextHolder.getContext().getAuthentication();
        if (auth != null && auth.getPrincipal() instanceof CustomUserDetails) {
            return (CustomUserDetails) auth.getPrincipal();
        }
        return null;
    }

    @PostMapping("/pedido/confirmar")
    @ResponseBody
    @SuppressWarnings("unchecked")
    public ResponseEntity<?> confirmar(@RequestBody Map<String, Object> payload, HttpSession session) {
        List<Map<String, Object>> items = (List<Map<String, Object>>) payload.getOrDefault("items", List.of());
        if (items.isEmpty()) {
            return ResponseEntity.unprocessableEntity().body(Map.of("error", "El carrito está vacío."));
        }

        Cliente cliente = clienteRepository.findByUsuario_CodigoUsuario(getUsuarioActual().getCodigoUsuario())
                .orElse(null);
        if (cliente == null) {
            return ResponseEntity.unprocessableEntity().body(Map.of("error", "No se encontró el perfil de cliente."));
        }

        Integer codigoRestaurante = pedidoDesdeCarritoService.resolverCodigoRestaurante(items);
        if (codigoRestaurante == null) {
            return ResponseEntity.unprocessableEntity().body(Map.of("error", "No se pudo determinar el restaurante."));
        }

        BigDecimal total = items.stream()
                .map(item -> {
                    BigDecimal precio = item.get("precio") != null
                            ? new BigDecimal(String.valueOf(item.get("precio"))) : BigDecimal.ZERO;
                    int cantidad = item.get("cantidad") != null ? Integer.parseInt(String.valueOf(item.get("cantidad"))) : 1;
                    return precio.multiply(BigDecimal.valueOf(cantidad));
                })
                .reduce(BigDecimal.ZERO, BigDecimal::add);

        try {
            Pago pago = new Pago();
            pago.setCliente(cliente);
            pago.setMonto(total);
            pago.setFechaPago(LocalDate.now());
            pago.setHoraPago(LocalTime.now());
            pago.setEstadoPago("Aceptado");

            Integer codigoPedido = pedidoDesdeCarritoService.crear(cliente, codigoRestaurante, items, pago);
            session.setAttribute("pedido_activo", codigoPedido);

            return ResponseEntity.ok(Map.of("redirect", "/rastreo"));
        } catch (Exception e) {
            return ResponseEntity.internalServerError().body(Map.of("error", "Error al procesar el pedido. Inténtalo nuevamente."));
        }
    }

    @GetMapping("/pedido/{id}/estado")
    @ResponseBody
    public ResponseEntity<?> estado(@PathVariable Integer id) {
        return pedidoRepository.findById(id)
                .map(pedido -> ResponseEntity.ok(Map.of("estado", pedido.getEstado())))
                .orElseGet(() -> ResponseEntity.status(404).body(Map.of("estado", "")));
    }

    @PostMapping("/pedido/{id}/cancelar")
    @ResponseBody
    public ResponseEntity<?> cancelar(@PathVariable Integer id) {
        Pedido pedido = pedidoRepository.findById(id).orElse(null);
        if (pedido == null) {
            return ResponseEntity.status(404).body(Map.of("error", "Pedido no encontrado"));
        }
        if (!"Iniciando".equals(pedido.getEstado())) {
            return ResponseEntity.unprocessableEntity().body(Map.of("error", "El pedido no puede cancelarse en su estado actual"));
        }

        pedido.setEstado("Cancelado");
        pedidoRepository.save(pedido);

        return ResponseEntity.ok(Map.of("ok", true));
    }
}
